pylops/optimization/solver.py

```diff
@@ -2,10 +2,202 @@
+        Estimated model of size :math:`[N \times 1]`
+    iit : :obj:`int`
+        Number of executed iterations
+    cost : :obj:`numpy.ndarray`, optional
+        History of the L2 norm of the residual
+
+    Notes
+    -----
+    Solve the :math:`\mathbf{y} = \mathbf{Opx}` problem using conjugate gradient
+    iterations [1]_.
+
+    .. [1] Hestenes, M R., Stiefel, E., “Methods of Conjugate Gradients for Solving
+       Linear Systems”, Journal of Research of the National Bureau of Standards.
+       vol. 49. 1952.
+
+    """
+
+    def __init__(self, Op, y):
+        self.Op = Op
+        self.y = y
+        self.ncp = get_array_module(y)
+
+    def _print_setup(self):
+        self.tstart = time.time()
+        if self.niter is None:
+            strpar = "tol = {self.tol:10e}\tniter = {self.niter}"
+        else:
+            strpar = "tol = {self.tol:10e}"
+        print(
+            "CG\n"
+            "-----------------------------------------------------------\n"
+            f"The Operator Op has {self.Op.shape[0]} rows and {self.Op.shape[1]} cols\n"
+        )
+        print(strpar)
+        print("-----------------------------------------------------------")
+        head1 = "    Itn           x[0]              r2norm"
+        print(head1)
+
+    def _print_step(self):
+        if not np.iscomplex(self.x[0]):
+            msg = f"{self.iiter:6g}        {self.x[0]:11.4e}        {self.cost[self.iiter]:11.4e}"
+        else:
+            msg = f"{self.iiter:6g}    {np.real(self.x[0]):4.1e} + {np.imag(self.x[0]):4.1e}j    {self.cost[self.iiter]:11.4e}"
+        print(msg)
+
+    def _print_finalize(self):
+        print(
+            f"\nIterations = {self.iiter}        Total time (s) = {time.time() - self.tstart:.2f}"
+        )
+        print("-----------------------------------------------------------------\n")
+
+    def setup(self, x0=None, niter=None, tol=1e-4, show=False):
+        """Setup solver
+
+        Parameters
+        ----------
+        x0 : :obj:`np.ndarray`, optional
+            Initial guess
+        niter : :obj:`int`, optional
+            Number of iterations
+        tol : :obj:`float`, optional
+            Tolerance on residual norm
+        show : :obj:`bool`, optional
+            Display setup log
+
+        """
+        self.tol = tol
+        self.niter = niter
+        if show:
+            self._print_setup()
+
+        # initialize solver
+        self.x0 = x0
+        if self.x0 is None:
+            self.x = self.ncp.zeros(self.Op.shape[1], dtype=self.y.dtype)
+            self.r = self.y.copy()
+        else:
+            self.x = self.x0.copy()
+            self.r = self.y - self.Op.matvec(self.x)
+        self.c = self.r.copy()
+        self.kold = self.ncp.abs(self.r.dot(self.r.conj()))
+
+        # create variables to track the residual norm and iterations
+        self.cost = []
+        self.cost.append(np.sqrt(self.kold))
+        self.iiter = 0
+
+    def step(self, show=False):

```

Comment:
I would not rely on properties here. A proper input and output is the clear way to go; it communicates to the user what is persistent information and what is used locally. Essentially I mean keep the same interface as the `solve()` method.

Reply:
I understand that you want to do `cg = CG(*args)` followed by `cg.setup()` and `cg.step()` to step-by-step debug your algorithm. I see that use-case, although it might not be a very common one (accept in development of the solver itself).

I see, this needs more careful though. I would probably design the steps a bit different. Perhaps

```
cg = CG(Op, x)
out = cg.solve(y)
```

I don't like that you are relying so much on properties. I have seen projects that rely on `self.xxx` being set, but then you want to add an extra feature down the road and it is not set anymore or incorrectly possibly messing up something else. As far as I am concerned the intermediate steps don't need to be saved as a property. I think it is a better design pattern. But, again, these are just some initial thoughts being blurted out late at night... I need to revisit them.

Reply:
Alright, I am a bit confused then. I wasn't so sure about class-based solvers (even though I myself opened a GitHub issue) but then when you suggested this I thought perhaps it is a good idea and we should give it a try.

But, if we simply design a class based solver to be 

```
cg = CG(Op, x)
out = cg.solve(y)
```

then I am not so sure I see the point. Can you explain where you see the benefit of using these 2 lines over the current `out=cg(Op,y,x)`?

My initial design was based on the idea that if we have a modular solver composed of setup-step-finalize a user could run one step at the time in a for loop and do anything else in between (similar to torch). On the other hand, a developer/interested user that wants to dig deeper in the solver working would likely need to focus only on the step method.

The point about relying on properties is a good one. Again, my initial idea was that to uniquely define a problem you must define:

- operator
- data
- solver

You could then run the solve method many times with different starting guesses and parameters. I definitely do not think it is good to initialize the solver with `x`, `y` makes much more sense (unless we want to only pass `Op` to begin with and pass `y` to `setup`. But then again we will rely on properties. However if we don't do that we will need to pass it to step and that sounds weird to me...

@cako, if you have time it would be good to have your opinion here too :)

Reply:
I am often faces with a different usecase: I want to solve the same problem for different kinds of data over and over again. So, in my view the data is not really "part of the problem". I could do something like

```
cg = CG(Op, x)
for i, y in enumerate(lots_of_y):
    out[i] = cg.solve(y)
```

Secondly, the main benefit is the encapsulation, code re-usability and templating (making it clearer for users what to define for new solvers). It allows for standardization of printing outputs and most likely makes things like switching to GPUs easier since all solvers are streamlined.

Reply:
Mmh ok, this depends on who you are and what you do... the other use case is also quite common, you want to solve the same problem with different starting guesses.

I would say we should try to make sure we accommodate for as many 'useful' use cases as possible. My new suggested flow is as follows.


In words:

- Create the solver object
- Setup the solver with data, solver hyperparameters (and, optionally, the initial guess). This will return the initial guess to be passed to step
- Step the solver Niter times passing x (at iter i-1) and obtaining x (at iter i)
- Optionally interleave step with callback
- Finalize solver

In code:

```
cgsolve = CG(Aop)

xhist = []
cgsolve.callback = lambda x: callback(x, xhist)

x = cgsolve.setup(y=y, x0=np.zeros(nx), tol=0, show=True)
for i in range(ny):
    x = cgsolve.step(x, show=True if i % 10 == 0 else False)
    cgsolve.callback(x)
cgsolve.finalize(show=True)
```

Reply:
Note that all our solvers are interchangeable between numpy (CPU) and copy (GPU), so the switching to GPU thing is already well taken into account and things won't change for the class based solvers

Reply:
Your example snippet is very long,I don't see a normal user writing that many lines. Everything that is not possible in two steps:

```
cg = CG(Op)
our = cg.solve(y, *args, **kwargs) 
```

More than that is not very user-friendly. It works for both multiple data and starting positions as well.

Reply:
> Also, note that all our solvers are interchangeable between numpy (CPU) and copy (GPU), so the switching to GPU thing is already well taken into account and things won't change for the class based solvers

Sure but the base class could call `get_array_module`, then it would be a bit more clear for a new user how to use it. But, agreed, not a mind-blowing change.

Reply:
Exactly, a common user will never do more that the two lines you write (or if they prefer and have to solve a single instance of the problem even use the associated function call). This was more meant to be what we do under the hood in `solve` or what an advanced user can do directly if they want to be flexible (eg do some plotting) and not do it within callback.

Regarding `get_array_module`, let me think about... maybe a better way is to create within the base class:

- setup
- solve
- finalize

which all have do something that will be in common for most solvers, and in each solver:

- _setup
- _solve
- _finalize

that we ask dev to implement when making a new solver and does the specific thing of that solver. An example

```
def setup(self, y=None, x0=None, show=False, **kwargs):
       self.y = y
        self.ncp = get_array_module(y)
        x = self._setup(x0, **kwargs)

        # print setup
        if show:
            self._print_setup(np.iscomplexobj(x))
        return x
```

and 

```
    def _setup(self, y=None, x0=None, niter=None, tol=1e-4, show=False):
 
        self.tol = tol
        self.niter = niter

        # initialize solver
        if x0 is None:
            x = self.ncp.zeros(self.Op.shape[1], dtype=self.y.dtype)
            self.r = self.y.copy()
        else:
            x = x0.copy()
            self.r = self.y - self.Op.matvec(x)
        self.c = self.r.copy()
        self.kold = self.ncp.abs(self.r.dot(self.r.conj()))

        # create variables to track the residual norm and iterations
        self.cost = []
        self.cost.append(np.sqrt(self.kold))
        self.iiter = 0
        
        return x
```

If this is more along the lines of what you had in mind (note this also makes the _print being called by the base class), I can do the same for the other methods?

Reply:
I tried this but didn't seem to add much value and a lot more verbosity. I think for now I'll stick with what I have